includes/class-wpcom-legacy-redirector-ui.php

```diff
@@ -0,0 +1,429 @@
+					$post_types = get_post_types();
+					$excerpt = get_the_excerpt( $post->ID );
+
+					// Check if redirect is a full URL or not.
+					if ( 0 === strpos( $excerpt, 'http' ) ) {
+						$redirect = $excerpt;
+					} elseif ( '/' === $excerpt ) {
+						$redirect = 'valid';
+					} else {
+						$redirect = home_url() . $excerpt;
+						if ( function_exists( 'wpcom_vip_get_page_by_path' ) ) {
+							$post_obj = wpcom_vip_get_page_by_path( $excerpt, OBJECT, $post_types );
+						} else {
+							$post_obj = get_page_by_path( $excerpt, OBJECT, $post_types );
+						}
+						if ( 404 !== $this->check_if_404( $redirect ) && ! is_null( $post_obj ) ) {
+							if ( 'publish' !== get_post_status( $post_obj->ID ) ) {
+								$redirect = 'private';
+							}
+						}
+					}
+				} else {
+					// If it's not stored as an Excerpt, it will be stored as a post_parent ID.
+					$post = get_post( $post->ID );
+					$parent = get_post( $post->post_parent );
+					$parent_post_status = get_post_status( $parent );
+
+					if ( is_null( get_post( $post->post_parent ) ) ) {
+						$redirect = 'null';
+					} elseif ( 'publish' !== get_post_status( $parent ) ) {
+						$redirect = 'private';
+					} else {
+						$parent_slug = $parent->post_name;
+						echo '/' . $parent_slug;
+						$redirect = home_url() . '/' . $parent_slug;
+					}
+				}
+				$status = $this->check_if_404( $redirect );
+
+				// Check if $redirect is invalid.
+				if ( ! wp_validate_redirect( $redirect, false ) ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => 'invalid',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+				// Check if $redirect is a 404.
+				if ( 404 === $status ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => '404',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+				// Check if $redirect is not publicly visible.
+				if ( 'private' === $redirect ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => 'private',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+				// Check if $redirect is pointing to a null Post ID.
+				if ( 'null' === $redirect ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => 'null',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+				// Check if $redirect is valid.
+				if ( wp_validate_redirect( $redirect, false ) && 404 !== $status || 'valid' === $redirect ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => 'valid',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+			}
+		}
+	}
+	public function generate_page_html() {
+		if ( ! current_user_can( 'manage_redirects' ) ) {
+			return;
+		}
+
+		$errors   = array();
+		$messages = array();
+		if ( class_exists( 'WPCOM_Legacy_Redirector' ) ) {
```

Comment:
How could this not exist? Could we check for this earlier on and if it's not there load it?